File: sdks/typescript/examples/test-secretstore.ts
/**
 * Secret Stores & Secrets Test
 *
 * Tests:
 *   1. Create a secret store
 *   2. List secret stores
 *   3. Get secret store by ID
 *   4. Update secret store
 *   5. Set secrets on a store
 *   6. List secrets (names only, values never returned)
 *   7. Create sandbox with secret store (inherits secrets)
 *   8. Verify secrets are injected as sealed env vars
 *   9. Delete secret
 *  10. Delete secret store
 *
 * Usage:
 *   npx tsx examples/test-projects.ts
 */

import { Sandbox, SecretStore } from "../src/index";

function green(msg: string) { console.log(`\x1b[32m\u2713 ${msg}\x1b[0m`); }
function red(msg: string) { console.log(`\x1b[31m\u2717 ${msg}\x1b[0m`); }
function bold(msg: string) { console.log(`\x1b[1m${msg}\x1b[0m`); }
function dim(msg: string) { console.log(`\x1b[2m  ${msg}\x1b[0m`); }

let passed = 0;
let failed = 0;

function check(desc: string, condition: boolean, detail?: string) {
  if (condition) {
    green(desc);
    passed++;
  } else {
    red(`${desc}${detail ? ` (${detail})` : ""}`);
    failed++;
  }
}

async function main() {
  bold("\n\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  bold("\u2551       Secret Stores & Secrets Test               \u2551");
  bold("\u255a\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255d\n");

  let storeId: string | null = null;
  let sandbox: Sandbox | null = null;
  const storeName = `test-store-${Date.now()}`;

  try {
    // -- Test 1: Create secret store --
    bold("--- Test 1: Create secret store ---\n");

    const store = await SecretStore.create({
      name: storeName,
      egressAllowlist: ["api.anthropic.com"],
    });

    storeId = store.id;
    check("Store created", !!store.id);
    check("Name matches", store.name === storeName);
    check("Egress allowlist set", store.egressAllowlist?.length === 1);
    dim(`Store ID: ${store.id}`);
    console.log();

    // -- Test 2: List secret stores --
    bold("--- Test 2: List secret stores ---\n");

    const stores = await SecretStore.list();
    check("List returns array", Array.isArray(stores));
    const found = stores.find((s) => s.id === storeId);
    check("Created store in list", !!found);
    dim(`Total stores: ${stores.length}`);
    console.log();

    // -- Test 3: Get secret store --
    bold("--- Test 3: Get secret store by ID ---\n");

    const fetched = await SecretStore.get(storeId!);
    check("Get returns correct store", fetched.id === storeId);
    check("Get has correct name", fetched.name === storeName);
    console.log();

    // -- Test 4: Update secret store --
    bold("--- Test 4: Update secret store ---\n");

    const updatedName = `${storeName}-updated`;
    const updated = await SecretStore.update(storeId!, {
      name: updatedName,
      egressAllowlist: ["api.anthropic.com", "*.openai.com"],
    });
    check("Name updated", updated.name === updatedName);
    check("Egress allowlist updated", updated.egressAllowlist?.length === 2);
    console.log();

    // -- Test 5: Set secrets --
    bold("--- Test 5: Set secrets ---\n");

    await SecretStore.setSecret(storeId!, "TEST_API_KEY", "sk-test-12345");
    green("Set TEST_API_KEY");

    await SecretStore.setSecret(storeId!, "DATABASE_URL", "postgres://localhost/test");
    green("Set DATABASE_URL");

    await SecretStore.setSecret(storeId!, "TEMP_SECRET", "will-be-deleted");
    green("Set TEMP_SECRET");
    console.log();

    // -- Test 6: List secrets --
    bold("--- Test 6: List secret entries ---\n");

    const entries = await SecretStore.listSecrets(storeId!);
    check("Returns array", Array.isArray(entries));
    const names = entries.map((e) => e.name);
    check("Has TEST_API_KEY", names.includes("TEST_API_KEY"));
    check("Has DATABASE_URL", names.includes("DATABASE_URL"));
    check("Has TEMP_SECRET", names.includes("TEMP_SECRET"));
    check("3 secrets total", entries.length === 3, `got ${entries.length}`);
    dim(`Secret names: ${names.join(", ")}`);
    console.log();

    // -- Test 7: Create sandbox with secret store --
    bold("--- Test 7: Create sandbox with secret store ---\n");

    sandbox = await Sandbox.create({
      secretStore: updatedName,
      timeout: 120,
    });
    check("Sandbox created", !!sandbox.sandboxId);
    dim(`Sandbox ID: ${sandbox.sandboxId}`);
    console.log();

    // -- Test 8: Verify secrets are sealed in sandbox --
    bold("--- Test 8: Verify secrets sealed in sandbox ---\n");

    // Secrets should be sealed tokens (osb_sealed_*) inside the VM.
    // The MITM proxy replaces sealed tokens with real values on outbound HTTPS requests,
    // so the real secret never exists in VM memory.
    const apiKeyResult = await sandbox.commands.run("echo $TEST_API_KEY");
    const apiKeyVal = apiKeyResult.stdout.trim();
    check("TEST_API_KEY is sealed", apiKeyVal.startsWith("osb_sealed_"),
      `got "${apiKeyVal}"`);

    const dbUrlResult = await sandbox.commands.run("echo $DATABASE_URL");
    const dbUrlVal = dbUrlResult.stdout.trim();
    check("DATABASE_URL is sealed", dbUrlVal.startsWith("osb_sealed_"),
      `got "${dbUrlVal}"`);

    const tempResult = await sandbox.commands.run("echo $TEMP_SECRET");
    const tempVal = tempResult.stdout.trim();
    check("TEMP_SECRET is sealed", tempVal.startsWith("osb_sealed_"),
      `got "${tempVal}"`);
    console.log();

    // -- Test 9: Delete secret --
    bold("--- Test 9: Delete secret ---\n");

    await SecretStore.deleteSecret(storeId!, "TEMP_SECRET");
    green("Deleted TEMP_SECRET");

    const afterDelete = await SecretStore.listSecrets(storeId!);
    const afterNames = afterDelete.map((e) => e.name);
    check("TEMP_SECRET removed", !afterNames.includes("TEMP_SECRET"));
    check("2 secrets remaining", afterDelete.length === 2, `got ${afterDelete.length}`);
    console.log();

    // -- Test 10: Delete secret store --
    bold("--- Test 10: Delete secret store ---\n");

    // Kill sandbox first
    await sandbox.kill();
    green("Sandbox killed");
    sandbox = null;

    await SecretStore.delete(storeId!);
    green("Secret store deleted");

    // Verify it's gone
    try {
      await SecretStore.get(storeId!);
      red("Store should not exist after delete");
      failed++;
    } catch {
      green("Store not found after delete (expected)");
      passed++;
    }
    storeId = null;
    console.log();

  } catch (err: any) {
    red(`Fatal error: ${err.message}`);
    if (err.stack) dim(err.stack);
    failed++;
  } finally {
    // Cleanup
    if (sandbox) {
      try { await sandbox.kill(); } catch {}
    }
    if (storeId) {
      try { await SecretStore.delete(storeId); } catch {}
    }
  }

  // --- Summary ---
  bold("========================================");
  bold(` Results: ${passed} passed, ${failed} failed`);
  bold("========================================\n");
  if (failed > 0) process.exit(1);
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});

File: sdks/python/examples/test_secretstore.py
#!/usr/bin/env python3
"""
Secret Stores & Secrets Test

Tests:
  1. Create a secret store
  2. List secret stores
  3. Get secret store by ID
  4. Update secret store
  5. Set secrets on a store
  6. List secrets (names only, values never returned)
  7. Create sandbox with secret store (inherits secrets)
  8. Verify secrets are injected as sealed env vars
  9. Delete secret
 10. Delete secret store

Usage:
  python examples/test_projects.py
"""

import asyncio
import sys
import time

from opencomputer import Sandbox, SecretStore

GREEN = "\033[32m"
RED = "\033[31m"
BOLD = "\033[1m"
DIM = "\033[2m"
RESET = "\033[0m"

passed = 0
failed = 0


def green(msg: str) -> None:
    print(f"{GREEN}\u2713 {msg}{RESET}")


def red(msg: str) -> None:
    print(f"{RED}\u2717 {msg}{RESET}")


def bold(msg: str) -> None:
    print(f"{BOLD}{msg}{RESET}")


def dim(msg: str) -> None:
    print(f"{DIM}  {msg}{RESET}")


def check(desc: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        green(desc)
        passed += 1
    else:
        red(f"{desc} ({detail})" if detail else desc)
        failed += 1


async def main() -> None:
    global passed, failed

    bold("\n\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557")
    bold("\u2551       Secret Stores & Secrets Test               \u2551")
    bold("\u255a\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255d\n")

    store_id = None
    sandbox = None
    store_name = f"test-store-{int(time.time())}"

    try:
        # -- Test 1: Create secret store --
        bold("--- Test 1: Create secret store ---\n")

        store = await SecretStore.create(
            name=store_name,
            egress_allowlist=["api.anthropic.com"],
        )

        store_id = store["id"]
        check("Store created", bool(store["id"]))
        check("Name matches", store["name"] == store_name)
        check("Egress allowlist set", len(store.get("egressAllowlist", [])) == 1)
        dim(f"Store ID: {store_id}")
        print()

        # -- Test 2: List secret stores --
        bold("--- Test 2: List secret stores ---\n")

        stores = await SecretStore.list()
        check("List returns list", isinstance(stores, list))
        found = any(s["id"] == store_id for s in stores)
        check("Created store in list", found)
        dim(f"Total stores: {len(stores)}")
        print()

        # -- Test 3: Get secret store --
        bold("--- Test 3: Get secret store by ID ---\n")

        fetched = await SecretStore.get(store_id)
        check("Get returns correct store", fetched["id"] == store_id)
        check("Get has correct name", fetched["name"] == store_name)
        print()

        # -- Test 4: Update secret store --
        bold("--- Test 4: Update secret store ---\n")

        updated_name = f"{store_name}-updated"
        updated = await SecretStore.update(
            store_id,
            name=updated_name,
            egress_allowlist=["api.anthropic.com", "*.openai.com"],
        )
        check("Name updated", updated["name"] == updated_name)
        check("Egress allowlist updated", len(updated.get("egressAllowlist", [])) == 2)
        print()

        # -- Test 5: Set secrets --
        bold("--- Test 5: Set secrets ---\n")

        await SecretStore.set_secret(store_id, "TEST_API_KEY", "sk-test-12345")
        green("Set TEST_API_KEY")

        await SecretStore.set_secret(store_id, "DATABASE_URL", "postgres://localhost/test")
        green("Set DATABASE_URL")

        await SecretStore.set_secret(store_id, "TEMP_SECRET", "will-be-deleted")
        green("Set TEMP_SECRET")
        print()

        # -- Test 6: List secrets --
        bold("--- Test 6: List secret entries ---\n")

        entries = await SecretStore.list_secrets(store_id)
        check("Returns list", isinstance(entries, list))
        names = [e["name"] for e in entries]
        check("Has TEST_API_KEY", "TEST_API_KEY" in names)
        check("Has DATABASE_URL", "DATABASE_URL" in names)
        check("Has TEMP_SECRET", "TEMP_SECRET" in names)
        check("3 secrets total", len(entries) == 3, f"got {len(entries)}")
        dim(f"Secret names: {', '.join(names)}")
        print()

        # -- Test 7: Create sandbox with secret store --
        bold("--- Test 7: Create sandbox with secret store ---\n")

        sandbox = await Sandbox.create(
            secret_store=updated_name,
            timeout=120,
        )
        check("Sandbox created", bool(sandbox.sandbox_id))
        dim(f"Sandbox ID: {sandbox.sandbox_id}")
        print()

        # -- Test 8: Verify secrets are sealed in sandbox --
        bold("--- Test 8: Verify secrets sealed in sandbox ---\n")

        # Secrets should be sealed tokens (osb_sealed_*) inside the VM.
        # The MITM proxy replaces sealed tokens with real values on outbound HTTPS requests,
        # so the real secret never exists in VM memory.
        api_key_result = await sandbox.commands.run("echo $TEST_API_KEY")
        api_key_val = api_key_result.stdout.strip()
        check("TEST_API_KEY is sealed", api_key_val.startswith("osb_sealed_"),
              f'got "{api_key_val}"')

        db_url_result = await sandbox.commands.run("echo $DATABASE_URL")
        db_url_val = db_url_result.stdout.strip()
        check("DATABASE_URL is sealed", db_url_val.startswith("osb_sealed_"),
              f'got "{db_url_val}"')

        temp_result = await sandbox.commands.run("echo $TEMP_SECRET")
        temp_val = temp_result.stdout.strip()
        check("TEMP_SECRET is sealed", temp_val.startswith("osb_sealed_"),
              f'got "{temp_val}"')
        print()

        # -- Test 9: Delete secret --
        bold("--- Test 9: Delete secret ---\n")

        await SecretStore.delete_secret(store_id, "TEMP_SECRET")
        green("Deleted TEMP_SECRET")

        after_delete = await SecretStore.list_secrets(store_id)
        after_names = [e["name"] for e in after_delete]
        check("TEMP_SECRET removed", "TEMP_SECRET" not in after_names)
        check("2 secrets remaining", len(after_delete) == 2, f"got {len(after_delete)}")
        print()

        # -- Test 10: Delete secret store --
        bold("--- Test 10: Delete secret store ---\n")

        # Kill sandbox first
        await sandbox.kill()
        green("Sandbox killed")
        sandbox = None

        await SecretStore.delete(store_id)
        green("Secret store deleted")

        # Verify it's gone
        try:
            await SecretStore.get(store_id)
            red("Store should not exist after delete")
            failed += 1
        except Exception:
            green("Store not found after delete (expected)")
            passed += 1
        store_id = None
        print()

    except Exception as e:
        red(f"Fatal error: {e}")
        import traceback
        traceback.print_exc()
        failed += 1
    finally:
        # Cleanup
        if sandbox:
            try:
                await sandbox.kill()
            except Exception:
                pass
        if store_id:
            try:
                await SecretStore.delete(store_id)
            except Exception:
                pass

    # --- Summary ---
    bold("========================================")
    bold(f" Results: {passed} passed, {failed} failed")
    bold("========================================\n")
    if failed > 0:
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())
